### PR TITLE
#184581359  User change password after x-time 

### PR DESCRIPTION
**Pull request by edwinnambaje**
#### What does this PR do?
The PR is for reminding users  to change their passwords every x amount of time 
since the time it was either manually updated.

#### Description of Task to be completed?
the PR has uses already existing endPoint
- PATCH /users/change-password

here the user after logging in he/she after a specified time for cronJob to run he/she gets an email that reminds him to change the password

#### How should this be manually tested?
after cloning the repo,
- in terminal add `npm install`
- add `.env` in your repo and assign the variables as there are in `.env.example` file
- include in `.env` PASSWORD_EXPIRATION_TIME
- run in terminal `npm run migrate:all` to setup the database
- run in terminal `npm start` to start the project
- After 3 minutes it will send an email that reminds you to update your password 
- in your browser add `http://localhost:<your port>/api-docs` for documentantion

#### Any background context you want to provide?
- Sendgrid helps to send email

#### What are the relevant pivotal tracker/Trello stories?
https://www.pivotaltracker.com/n/projects/2632830/stories/184581359

#### Screenshots (if appropriate)
Screenshot 2023-03-17 at 11 28 12
Screenshot 2023-03-17 at 12 31 43
Screenshot 2023-03-17 at 11 29 11

#### Questions:
n/a

### CODE REVIEW

**Review by Patienceineza — APPROVED** [review]
LGTM

File: src/helpers/eventEmitter.js [hunk 1]
```diff
@@ -0,0 +1,10 @@
+const EventEmitter = require('events');
+
+const eventEmitter = new EventEmitter();
+
+// Registering a listener for the 'passwordUpdated' event
+eventEmitter.on('passwordUpdated', (userId) => {
+  console.log(`Password has been updated for user with ID ${userId}`);
```
Comment (nshutijonathan) [inline src/helpers/eventEmitter.js]: remove console.log,it is for debugging purpose

File: src/middlewares/passwordExpiration.js [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+};
+
+// eslint-disable-next-line arrow-body-style
+const hasUserIgnoredPasswordUpdatePrompt = (user) => {
+  return user.mustUpdatePassword === true && hasPasswordExpired(user);
+};
+
+const checkPasswordExpirationCronJob = new CronJob('0 */3 * * * *', async () => {
+  try {
+    console.log('Checking for password expiration...');
```
Comment (nshutijonathan) [inline src/middlewares/passwordExpiration.js]: remove console.log

File: src/middlewares/passwordExpiration.js [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+
+    const usersk = await db.users.findAll();
+
+    usersk.forEach(async (user) => {
+      if (hasPasswordExpired(user)) {
+        sendPasswordUpdateEmail(user);
+
+        promptPasswordUpdate(user);
+      } else if (hasUserIgnoredPasswordUpdatePrompt(user)) {
+        console.log(`${user.username} ignored password update prompt`);
```
Comment (nshutijonathan) [inline src/middlewares/passwordExpiration.js]: remove console.log

File: src/middlewares/passwordExpiration.js [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+        updateUserMustUpdatePasswordField(user, false);
+      }
+    });
+  } catch (error) {
+    console.error(`Error checking password expiration: ${error}`);
```
Comment (nshutijonathan) [inline src/middlewares/passwordExpiration.js]: remove console.log

**Review by nshutijonathan — CHANGES_REQUESTED** [review]
Please update reame and remove the console.log

**Review by Zeden12 — APPROVED** [review]
LGTM

**Review by Martha-Iradukunda — APPROVED** [review]
LGTM

File: src/helpers/mailTemplate.js [hunk 5]
```diff
@@ -12,7 +12,16 @@ export const verifyEmailTemplate = (userToken) => {
     </div>
     `;
 };
-
+export const changePasswordTemplate = (user, passwordExpirationTime) => {
+  return `
+    <div style="background-color: #f2f2f2; padding: 20px;">
+      <h1 style="color: #004d99; text-align: center;">Welcome to Brogrammers e-commerce dear!</h1>
+      <p style="color: #000; font-size: 16px;">Dear ${user.username},\n\nYour password will expire in less than ${passwordExpirationTime} minutes.. Please use the following link to change your password:</p>
+      <a href="${process.env.FRONTEND_URL}/users/change-password" style="display: block; text-align: center; padding: 10px 20px; background-color: #004d99; color: #fff; text-decoration: none; border-radius: 5px; margin-top: 20px;">Change Password</a>
```
Comment (UmuhireJessie) [inline src/helpers/mailTemplate.js:20]: You should add (FRONTEND_URL) environment variable in your `.env.example`

File: src/helpers/eventEmitter.js (hunk 1 above)
Comment (edwinnambaje) [reply src/helpers/eventEmitter.js]: done

File: src/middlewares/passwordExpiration.js (hunk 2 above)
Comment (edwinnambaje) [reply src/middlewares/passwordExpiration.js]: done

File: src/middlewares/passwordExpiration.js (hunk 4 above)
Comment (edwinnambaje) [reply src/middlewares/passwordExpiration.js]: done

File: src/middlewares/passwordExpiration.js [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+const promptPasswordUpdate = (user) => {
+  updateUserMustUpdatePasswordField(user, true);
+};
+
+// eslint-disable-next-line arrow-body-style
+const hasUserIgnoredPasswordUpdatePrompt = (user) => {
+  return user.mustUpdatePassword === true && hasPasswordExpired(user);
+};
+
+const checkPasswordExpirationCronJob = new CronJob('0 0 */2 * * *', async () => {
```
Comment (UmuhireJessie) [inline src/middlewares/passwordExpiration.js]: Prefer using an environment variable. A client might want the job to run after weeks, days or even months

**Review by UmuhireJessie — COMMENTED** [review]
Good job Edwin!
Two things:

- add FRONTEND_URL and PASSWORD_EXPIRATION_TIME in environment variable in your `.env.example`
- Prefer to use an environment variable to specify after how much time the job will run

File: tests/users.test.js [hunk 7]
```diff
@@ -148,9 +160,10 @@ describe('testing user profile', () => {
       .set({ authorization: `bearer ${token}` })
       .send({
         gender: 'female',
-        email: 'edwin12@gmail.com',
+        email: 'lucy@gmail.com'
       })
       .end((error, res) => {
+        console.log(res.body);
```
Comment (nshutijonathan) [inline tests/users.test.js]: remove console.log

**Review by nshutijonathan — COMMENTED** [review]
remove all console.log

File: src/middlewares/passwordExpiration.js [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+      if (hasPasswordExpired(user)) {
+        sendPasswordUpdateEmail(user);
+
+        promptPasswordUpdate(user);
+      } else if (hasUserIgnoredPasswordUpdatePrompt(user)) {
+        updateUserMustUpdatePasswordField(user, false);
+      }
+    });
+  } catch (error) {
+    console.error(`Error checking password expiration: ${error}`);
```
Comment (nshutijonathan) [inline src/middlewares/passwordExpiration.js]: you still have console.log

Comment (edwinnambaje) [reply src/middlewares/passwordExpiration.js]: solved now

**Review by INValentin — APPROVED** [review]
LGTM

**Review by nshutijonathan — APPROVED** [review]
LGTM

**Review by blackd44 — APPROVED** [review]
looks good

**Review by UmuhireJessie — APPROVED** [review]
Good!

**Pull request merged**